src/hal/src/command/raw.rs

```diff
@@ -441,19 +496,23 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
         stride: u32,
     );
 
-    ///
+    /// Begins a query operation.  Queries count operations or record timestamps
+    /// resulting from commands that occur between the beginning and end of the query,
+    /// and save the results to the query pool.
     fn begin_query(&mut self, query: Query<B>, flags: QueryControl);
 
-    ///
+    /// End a query.
     fn end_query(&mut self, query: Query<B>);
 
-    ///
+    /// Reset/clear the values in the given range of the query pool.
     fn reset_query_pool(&mut self, pool: &B::QueryPool, queries: Range<QueryId>);
 
-    ///
+    /// Requests a timestamp to be written.
     fn write_timestamp(&mut self, pso::PipelineStage, Query<B>);
 
-    ///
+    /// Modify constant data in a graphics pipeline.
+    /// Push constants are intended to modify data in a pipeline more
```

Comment:
data in the pipeline layout (AMD uses the term `user data`). It's basically a mem copy, the alternative would be updating the values inside a descriptor set. The better performance stems from the addtional indirection due to the descriptor sets setting inside the 'user data' and the shader has to dereference the descriptor set pointer